yt/chyt/server/query_analyzer.cpp

```diff
@@ -728,6 +744,125 @@ void TQueryAnalyzer::OptimizeQueryProcessingStage()
+        }
+
+        // All requested directions must align and form either a forward or backward read of the underlying table.
+        auto columnDirection = GetReadInOrderMode(*schemaColumn.SortOrder(), orderByElement->direction);
+        if (commonDirection == EReadInOrderMode::None) {
+            commonDirection = columnDirection;
+        } else if (commonDirection != columnDirection) {
+            return EReadInOrderMode::None;
+        }
+
+        // See float-related comments below.
+        if (IsFloatingPointType(schemaColumn.CastToV1Type()) && !orderByElement->nulls_direction_was_explicitly_specified) {
+            floatKeyColumnsWithoutExplicitNullsDirection.push_back(schemaColumn.Name());
+        }
+    }
+
+    // See the last comment in this function.
+    if (!floatKeyColumnsWithoutExplicitNullsDirection.empty()) {
+        THROW_ERROR_EXCEPTION(
+            "Nulls direction must be specified explicitly for key columns of floating-point "
+            "types when using optimize_read_in_order is enabled; see docs for more details")
+            << TErrorAttribute("problematic_columns", floatKeyColumnsWithoutExplicitNullsDirection);
+    }
+
+    // In YT, NULL values compare less than any other values. In ClickHouse, NULL value placement
+    // depends solely on the placement requested in the query, which defaults to NULLS LAST.
+    // Since we combine both YT and CH sorting in this mode, there is only one acceptable direction
+    // depending on the requested sort order. In cases when the unsupported option turns out to be
+    // the default (for ascending sorts), we tweak the query to avoid additional user turmoil.
+    if (commonDirection != EReadInOrderMode::None) {
+        for (const auto& orderByElementAst : selectQuery->orderBy()->children) {
+            auto orderByElement = orderByElementAst->as<DB::ASTOrderByElement>();
+
+            // The value of nulls_direction is equal to direction for NULLS LAST and to the opposite
+            // of direction for NULLS FIRST.
+            // For ASC (d = 1), we need NULLS FIRST (nd = -d = -1).
+            // For DSC (d = -1), we need NULLS LAST (nd =  d = -1).
+            if (orderByElement->nulls_direction != -1) {
+                THROW_ERROR_EXCEPTION_IF(
+                    orderByElement->nulls_direction_was_explicitly_specified,
+                    "Specified nulls direction is incompatible with enabled optimize_read_in_order mode; "
+                    "please turn it off via query settings, specify another nulls direction or use the default nulls direction");
+                YT_LOG_DEBUG(
+                    "Reversing ORDER BY nulls direction (ColumnName: %v, Direction: %v, NullsDirection: %v -> -1)",
+                    orderByElement->children.front()->getColumnName(),
+                    orderByElement->direction,
+                    orderByElement->nulls_direction);
+                orderByElement->nulls_direction = -1;
```

Comment:
Looks suspicious.
First, I'm not sure if we are allowed to change the query AST. I have no idea how CH code would handle this "unexpected change".
Second, it could change the query results compared to the execution without the optimization.

The second point would prevent this optimization from being turned on by default. If this behavior is desired, we could introduce a separate option to configure it.

Reply:
For the first point, I don't know how to check whether it is okay to change it like that or not. I can say that first I only changed `nulls_direction` and it didn't have any effect, but once you also set `nulls_direction_was_explicitly_set` it starts having effect. There are tests for this and they pass. 
Also there are already some AST changes nearby in `Analyze`, so I think it should be fine...

Reply:
As for the second point, my intent was exactly so that we could enable it by default and the users wouldn't have to manually write `NULLS FIRST` or `NULLS LAST` in their queries. 
I agree that this is all quite annoying. I am also concerned that optional key columns are probably not uncommon, since that is pretty much what you get by default.

I can introduce a separate option that would enable this behaviour and turn it off by default. If it is disabled, then we will throw an exception here instead of changing the AST. However, I still want to change the AST, because enabling this extra option is much easier for the user than thinking about the correct NULLS direction. It is also much easier to copy and paste an enabled option from query to query. 

What do you think?

Reply:
I thought about it one more time and now I'm pretty sure that this type of change is illegal.

For example, let's look at this query:

```sql
select * from (
    select * from "//sorted_table"
    order by key
    limit 5
    union all
    select * from "//unsorted_table"
    order by key
    limit 5
) order by key
```

In this case ClickHouse may substitute the last `order by key` phase with `sorted merge`, because both subqueries have already been sorted by the key.

But if you change the order of nulls, two subqueries would be sorted differently, and sorted merge may either throw an error or return an incorrect result.

The same applies to NaNs - I'm not sure that it's a good idea to have this logic at all, even under an option.

Reply:
The problem is that without this logic we might get incorrect results when read_in_order is enabled. What should we do?

Reply:
New approach: we do not change AST in any way, but add two additional options `assume_no_null_keys` and `assume_no_nan_keys`.
We check nulls direction for a key column if it could have either nulls or nans, but not both at the same time (only for floats), and only optimize if it is correct. If a float key column could have both nulls and nans we never optimize.